### PR TITLE
Role System: Adding dynamic polymorphic role system

### PR DESCRIPTION
**Pull request by Alf21**

### CODE REVIEW

File: code/player/Player.RPCs.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+using Sandbox;
+
+using TTTReborn.Roles;
+
+namespace TTTReborn.Player
+{
+    partial class TTTPlayer
+    {
+        /// <summary>
+        /// Called by the server, updates client's `Role`.
```
Comment (mzegar) [inline code/player/Player.RPCs.cs:10]: Very nit picky...

"Must be called on the server, updates TTTPlayer's `Role`.

File: code/player/Player.cs [hunk 2]
```diff
@@ -2,16 +2,30 @@
...
+                {
+                    SetRole(To.Single(this), role.Name);
+                }
+            }
+            get {
+                return role;
+            }
+        }
+
+        private BaseRole role = new NoneRole();
```
Comment (mzegar) [inline code/player/Player.cs:28]: Unless I'm misunderstanding, where do we keep track of the players role on the server?

**Review by mzegar — COMMENTED** [review]
Quick question

**Comment by TimGoll** [comment]
Just a general question. I personally prefer naming variables and classes with the "type" in the beginning. So instead of "NoneRole" it would be "RoleNone". What's the general opinion on this?

File: code/player/Player.RPCs.cs (hunk 1 above)
Comment (Alf21) [reply code/player/Player.RPCs.cs:10]: Hmm. Good idea regarding the missing separation between Client and Server methods.

File: code/player/Player.cs (hunk 2 above)
Comment (Alf21) [reply code/player/Player.cs:28]: Currently, the role on the server is the role the client has. No client is aware of the role of other players, but I thought that's something for another PR as it gonna be complex (storing information the client is aware of, displaying other roles to the clients than they actually have, etc.).

**Comment by Alf21** [comment]
> Just a general question. I personally prefer naming variables and classes with the "type" in the beginning. So instead of "NoneRole" it would be "RoleNone". What's the general opinion on this?

hmmm,... You don't say ArrayList ArrayListPlayer. Most of all, you're saying playerList, right? ^^

**Comment by mzegar** [comment]
> > Just a general question. I personally prefer naming variables and classes with the "type" in the beginning. So instead of "NoneRole" it would be "RoleNone". What's the general opinion on this?
> 
> hmmm,... You don't say ArrayList ArrayListPlayer. Most of all, you're saying playerList, right? ^^

In the case of `Role`, I personally think `TraitorRole` rolls off the tongue a bit better.

**Comment by Alf21** [comment]
> > > Just a general question. I personally prefer naming variables and classes with the "type" in the beginning. So instead of "NoneRole" it would be "RoleNone". What's the general opinion on this?
> 
> > 
> 
> > hmmm,... You don't say ArrayList ArrayListPlayer. Most of all, you're saying playerList, right? ^^
> 
> 
> 
> In the case of `Role`, I personally think `TraitorRole` rolls off the tongue a bit better. We should however typically avoid the type in the name where possible.

Then how about just "Traitor" and "Innocent"? ^^

**Review by mzegar — APPROVED** [review]
LGTM, I'm fine with just "Traitor" tbh.

**Pull request merged**